crates/ruff_linter/src/rules/fastapi/rules/fastapi_unused_path_parameter.rs

```diff
@@ -0,0 +1,222 @@
+/// ## Why is this bad?
+/// Path parameters are used to extract values from the URL path.
+/// If a path parameter is declared in the route path but not in the function signature, it will not be accessible in the function body, which is likely a mistake.
+///
+/// ## Known problems
+/// If the path parameter is not a valid Python identifier, FastAPI will normalize it to a valid identifier.
+/// This lint simply ignores path parameters that are not valid identifiers, as that normalization behavior is undocumented.
+///
+/// ## Example
+///
+/// ```python
+/// from fastapi import FastAPI
+///
+/// app = FastAPI()
+///
+///
+/// @app.get("/things/{thing_id}")
+/// async def read_thing(query: str):
+///     ...
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// from fastapi import FastAPI
+///
+/// app = FastAPI()
+///
+///
+/// @app.get("/things/{thing_id}")
+/// async def read_thing(thing_id: int, query: str):
+///     ...
+/// ```
+
+#[violation]
+pub struct FastApiUnusedPathParameter {
+    arg_name: String,
+    function_name: String,
+    arg_name_already_used: bool,
+}
+
+impl Violation for FastApiUnusedPathParameter {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self {
+            arg_name,
+            function_name,
+            arg_name_already_used: arg_name_is_pos_only_arg,
+        } = self;
+        if *arg_name_is_pos_only_arg {
+            format!(
+                "Path parameter `{arg_name}` in route path, but appears as a positional-only argument in function `{function_name}`. Consider making it a regular argument."
+            )
+        } else {
+            format!(
+                "Path parameter `{arg_name}` in route path but not in function `{function_name}`."
+            )
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let Self { arg_name, .. } = self;
+        if self.arg_name_already_used {
+            None
+        } else {
+            Some(format!("Add `{arg_name}` to function signature"))
+        }
+    }
+}
+
+struct PathParamIterator<'a> {
+    input: &'a str,
+    chars: Peekable<CharIndices<'a>>,
+}
+
+impl<'a> PathParamIterator<'a> {
+    fn new(input: &'a str) -> Self {
+        PathParamIterator {
+            input,
+            chars: input.char_indices().peekable(),
+        }
+    }
+}
+
+impl<'a> Iterator for PathParamIterator<'a> {
+    type Item = (&'a str, Range<usize>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some((start, c)) = self.chars.next() {
+            if c == '{' {
+                if let Some((end, _)) = self.chars.by_ref().find(|&(_, ch)| ch == '}') {
+                    let param_content = &self.input[start + 1..end];
+                    // We ignore text after a colon, since those are path convertors
+                    // See also: https://fastapi.tiangolo.com/tutorial/path-params/?h=path#path-convertor
```

Comment:
Since path parameters use the same format as f-strings, could we use our parser here instead? Just parse the path, extract the f-string segments? You can see [here](https://play.ruff.rs/8847ec0d-dcd0-417d-ba75-b372b96a3115) that its evaluated to a literal element, followed by an `FStringExpressionElement` where the `:` part is the `FStringFormatSpec`.

Reply:
Path parameters can use invalid identifier names. For example, `/route/{path-name}`, which the f-string parser doesn't support. However, we just ignore invalid identifiers in the rule anyways since fastapi normalizes them in an undocumented way, so i don't think this is a blocker to using that parser.
Also, fastapi doesn't handle the `{name!r}` and `{name=}` special syntaxes and would normalize them as `name_r` and `name` respectively, but this is a very obscure edge-case so it doesn't matter much. We do need to at least disable the rule if there is a conversion or debug_text in the f-string to avoid complaining about the wrong name though.
If those aren't a problem, i'll switch to the f-string parser

Reply:
I wrote the code, but it required to switch back to using String instead of &str, not too sure if that's an issue

Reply:
I leave this up to you @charliermarsh but cloning the AST, then parsing it seems very heavy weight.

Reply:
Ok yeah, let's just skip the parser idea for now then.

Does FastAPI allow you to escape curly braces? Like `/foo{{bar}}/` to represent a literal `{` and `}` in the path, as with f-strings?

Reply:
it doesn't, it will just be treated as if you had a path attribute named `bar` (without curly braces).
Although i'm pretty sure it's because of the aforementioned undocumented normalizing behavior.